modulus/models/diffusion/song_unet.py

```diff
@@ -364,3 +364,264 @@ def forward(self, x, noise_labels, class_labels, augment_labels=None):
+                torch.permute(
+                    torch.reshape(selected_pos_embd, (self.pos_embd.shape[0], B, X, Y)),
+                    (1, 0, 2, 3),
+                )
+                .to(x.device)
+                .to(x.dtype)
+            )  # (B, N_pe, X, Y)
+        x = torch.cat((x, selected_pos_embd), dim=1)
+        if self.embedding_type != "zero":
+            # Mapping.
+            emb = self.map_noise(noise_labels)
+            emb = (
+                emb.reshape(emb.shape[0], 2, -1).flip(1).reshape(*emb.shape)
+            )  # swap sin/cos
+            if self.map_label is not None:
+                tmp = class_labels
+                if self.training and self.label_dropout:
+                    tmp = tmp * (
+                        torch.rand([x.shape[0], 1], device=x.device)
+                        >= self.label_dropout
+                    ).to(tmp.dtype)
+                emb = emb + self.map_label(tmp * np.sqrt(self.map_label.in_features))
+            if self.map_augment is not None and augment_labels is not None:
+                emb = emb + self.map_augment(augment_labels)
+            emb = silu(self.map_layer0(emb))
+            emb = silu(self.map_layer1(emb))
+        else:
+            emb = torch.zeros(
+                (noise_labels.shape[0], self.emb_channels), device=x.device
+            )
+
+        # Encoder.
+        skips = []
+        aux = x
+        for name, block in self.enc.items():
+            with nvtx.annotate(f"SongUNet encoder: {name}", color="blue"):
+                if "aux_down" in name:
+                    aux = block(aux)
+                elif "aux_skip" in name:
+                    x = skips[-1] = x + block(aux)
+                elif "aux_residual" in name:
+                    x = skips[-1] = aux = (x + block(aux)) / np.sqrt(2)
+                else:
+                    x = block(x, emb) if isinstance(block, UNetBlock) else block(x)
+                    skips.append(x)
+
+        # Decoder.
+        aux = None
+        tmp = None
+        for name, block in self.dec.items():
+            with nvtx.annotate(f"SongUNet decoder: {name}", color="blue"):
+                if "aux_up" in name:
+                    aux = block(aux)
+                elif "aux_norm" in name:
+                    tmp = block(x)
+                elif "aux_conv" in name:
+                    tmp = block(silu(tmp))
+                    aux = tmp if aux is None else tmp + aux
+                else:
+                    if x.shape[1] != block.in_channels:
+                        x = torch.cat([x, skips.pop()], dim=1)
+                    x = block(x, emb)
+        return aux
+
+    def positionalEmbedding(self):
+        if self.N_grid_channels == 0:
+            return None
+        elif self.gridtype == "learnable":
+            grid = torch.nn.Parameter(
+                torch.randn(
+                    self.N_grid_channels, self.img_resolution, self.img_resolution
+                )
+            )
```

Comment:
Thinking forward to the ICON hpx work. It would be more flexible to have the position embedding be a 2D array with shape (channels, n_global_grid), and `global_index` be a (batch, patch, patch) array of indices into this array.

I worry we'll need to have another backwards incompatible change to support the HPX work without this change. Is it hard?

Reply:
agreed to defer this to future work.